### PR TITLE
Start working on removing global JS IPython

### PR DESCRIPTION
**Pull request by Carreau**
1) rename to jupyter.
2) Start removing this weird "we require the global namespace to add ourself to it"

Note that 2 is all or nothing, or requireJS complain that `module Xyz has not yet be loaded in current contex`.

So if we agreed to trop the IPython/Jupyter global (at least to store module) for 4.0 that would be great and simplify things a lot.

### CODE REVIEW

**Comment by Carreau** [comment]
This now uses `properties` to load lazily the module/constructor and console.log a deprecation warning.
So require does not complain.

**Comment by minrk** [comment]
Are we leaving the global `IPython` for backward compat, or no?

**Comment by takluyver** [comment]
I think we should be leaving the IPython name for a while. I know we say the JS API isn't stable, but people are using it anyway, and breaking all code when it's easy enough to leave that reference around would be a bit rude.

**Comment by damianavila** [comment]
> I think we should be leaving the IPython name for a while. I know we say the JS API isn't stable, but people are using it anyway, and breaking all code when it's easy enough to leave that reference around would be a bit rude.

I completely agree with this opinion...

**Comment by Carreau** [comment]
The global is also super useful for debugging, but I'm for slowly deprecating it with a message.

Here it's how it appears as soon as I type dot : 
screen shot 2015-05-21 at 10 10 21

**Comment by takluyver** [comment]
Does it do the same with instances in the `IPython` namespace, or only with classes? `require()` doesn't help you get instances.

**Comment by Carreau** [comment]
Right now only with Classes, most of the instances should be attached to the object that created them IMHO, but this is another discussion. 

I have no opposition (for now) of `main.js` mutating the global namespace, I just don't want the leaves of the dependency tree to require a global and mutate it.

**Comment by Carreau** [comment]
Moar commits. less globals.

**Comment by jdfreder** [comment]
Along the lines of what @minrk was saying in today's meeting, you still need a way to expose the singleton instances, which are true globals.  Looking at this PR, it looks like you aren't doing that (correct me if I'm wrong).  How about making the instances require-able via main.js?  i.e. `IPython.notebook` -> `require('notebook/js/main').notebook`?

**Comment by Carreau** [comment]
I don't delete the IPython, neither remove the singleton instance. They are still exposed, I won't remove them in this PR.

**Comment by Carreau** [comment]
Removed `[WIP]`. 
I start to reach the maximum of things that can be removed easily, and longer will be hard for review.

**Comment by Carreau** [comment]
Travis is happy.

**Comment by Carreau** [comment]
Ping.

File: notebook/static/base/js/security.js [hunk 1]
```diff
@@ -59,7 +57,7 @@ define([
          * sanitize just the css in style tags in a block of html
          * called by sanitize_html, if allow_css is true
          */
-        var h = $("<div/>").append(html);
+        var h = document.createElement('div').append(html);
```
Comment (minrk) [inline notebook/static/base/js/security.js]: I don't think this works. element.append only adds a single node, not hierarchical html text.

Comment (minrk) [inline notebook/static/base/js/security.js]: Plus, `h.html()` called below won't work either.

Comment (Carreau) [inline notebook/static/base/js/security.js]: reverted

**Comment by jdfreder** [comment]
You'll want to update `examples/Custom Keyboard shortcuts.ipynb` to use Jupyter instead of IPython.  I noticed this while working on #124 , but since this isn't merged, I didn't do it there.

**Comment by Carreau** [comment]
@jdfreder Done, thanks.  I also updated the instructions a bit.

File: examples/Notebook/Custom Keyboard Shortcuts.ipynb [hunk 2]
```diff
@@ -38,6 +38,13 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
+    "The keypress `r` in command mode is by default bound to change the type of a cell ro `raw`. This action will thus be unavaillable from the keyboard now."
```
Comment (jdfreder) [inline examples/Notebook/Custom Keyboard Shortcuts.ipynb]: cell **to** `raw`

Also, I think moving things around in the sentence makes it read better:
"By default the keypress `r`, while in command mode, changes the type of the selected cell to `raw`.  This shortcut is overridden by the code in the previous cell, and thus the action no longer be available via the keypress `r`."

File: notebook/static/base/js/namespace.js [hunk 3]
```diff
@@ -1,10 +1,86 @@
 // Copyright (c) Jupyter Development Team.
 // Distributed under the terms of the Modified BSD License.
+//
+//
```
Comment (jdfreder) [inline notebook/static/base/js/namespace.js]: Accident?

File: notebook/static/base/js/namespace.js [hunk 4]
```diff
@@ -1,10 +1,86 @@
...
 
-var IPython = IPython || {};
-define([], function(){
+
+var Jupyter = Jupyter || {};
+
+var jprop = function(name, module_path){
+    Object.defineProperty(Jupyter, name, {
+      get: function() { 
+          console.warn('accessing `'+name+'` is deprecated. Use require("'+module_path+'")');
```
Comment (jdfreder) [inline notebook/static/base/js/namespace.js]: You surround `name` in backticks, so you should surround `require("...")` in backticks too.

File: notebook/static/base/js/namespace.js [hunk 5]
```diff
@@ -1,10 +1,86 @@
...
+      },
+      enumerable: true,
+      configurable: false
+    });
+}
+
+var jglobal = function(name, module_path){
+    Object.defineProperty(Jupyter, name, {
+      get: function() { 
+          console.warn('accessing `'+name+'` is deprecated. Use require("'+module_path+'").'+name);
```
Comment (jdfreder) [inline notebook/static/base/js/namespace.js]: Same here : )

**Comment by jdfreder** [comment]
Few comments, otherwise this has my stamp of approval !  :ok:

File: examples/Notebook/Custom Keyboard Shortcuts.ipynb (hunk 2 above)
Comment (Carreau) [inline examples/Notebook/Custom Keyboard Shortcuts.ipynb]: done.

File: notebook/static/base/js/namespace.js (hunk 3 above)
Comment (Carreau) [inline notebook/static/base/js/namespace.js]: done

File: notebook/static/base/js/namespace.js (hunk 4 above)
Comment (Carreau) [inline notebook/static/base/js/namespace.js]: done

File: notebook/static/base/js/namespace.js (hunk 5 above)
Comment (Carreau) [inline notebook/static/base/js/namespace.js]: and odne

**Comment by jdfreder** [comment]
Thanks @Carreau

**Pull request merged**